src/pybind/mgr/volumes/module.py

```diff
@@ -0,0 +1,430 @@
+        return self._volume_get_fs(vol_name) is not None
+
+    def _cmd_fs_subvolume_create(self, inbuf, cmd):
+        vol_name = cmd['vol_name']
+        sub_name = cmd['sub_name']
+
+        size = cmd.get('size', None)
+
+        if not self._volume_exists(vol_name):
+            return -errno.ENOENT, "", \
+                   "Volume not found, create it with `ceph volume create` " \
+                   "before trying to create subvolumes"
+
+        # TODO: validate that subvol size fits in volume size
+
+        with CephFSVolumeClient(rados=self.rados) as vc:
+            # TODO: support real subvolume groups rather than just
+            # always having them 1:1 with subvolumes.
+            vp = VolumePath(sub_name, sub_name)
+
+            vc.create_volume(vp, size)
+
+        return 0, "", ""
+
+    def _cmd_fs_subvolume_delete(self, inbuf, cmd):
+        vol_name = cmd['vol_name']
+        sub_name = cmd['sub_name']
+
+        fs = self._volume_get_fs(vol_name)
+        if fs is None:
+            return 0, "", "Volume '{0}' already deleted".forma(vol_name)
+
+        vol_fscid = fs['id']
+
+        with CephFSVolumeClient(rados=self.rados) as vc:
+            # TODO: support real subvolume groups rather than just
+            # always having them 1:1 with subvolumes.
+            vp = VolumePath(sub_name, sub_name)
+
+            vc.delete_volume(vp)
+
+        # TODO: create a progress event
+        self._background_jobs.put(PurgeJob(vol_fscid, vp))
+
+        return 0, "", ""
+
+    def _cmd_fs_volume_delete(self, inbuf, cmd):
+        vol_name = cmd['vol_name']
+
+        # Tear down MDS daemons
+        # =====================
+        spec = orchestrator.StatelessServiceSpec()
+        spec.name = vol_name
+        try:
+            completion = self._oremote(
+                "rm_stateless_service",
+                "mds",
+                spec
+            )
+            self._wait([completion])
+        except ImportError:
+            self.log.warning("No orchestrator, not tearing down MDS daemons")
+        except Exception as e:
+            # Don't let detailed orchestrator exceptions (python backtraces)
+            # bubble out to the user
+            self.log.exception("Failed to tear down MDS daemons")
+            return -errno.EINVAL, "", str(e)
+
+        if self._volume_exists(vol_name):
+            # In case orchestrator didn't tear down MDS daemons cleanly, or
+            # there was no orchestrator, we force the daemons down.
+            r, out, err = self.mon_command({
+                'prefix': 'fs set',
+                'fs_name': vol_name,
+                'var': 'cluster_down',
+                'val': 'true'
+            })
+            if r != 0:
+                return r, out, err
+
+            for mds_name in self._volume_get_mds_daemon_names(vol_name):
+                r, out, err = self.mon_command({
+                    'prefix': 'mds fail',
+                    'role_or_gid': mds_name})
```

Comment:
Since you have the information, I think it'd be cleaner to use `fscid:rank` instead.